tensorboard/backend/application.py

```diff
@@ -328,7 +336,8 @@ def __call__(self, environ, start_response):  # pylint: disable=invalid-name
     """Central entry point for the TensorBoard application.
 
     This method handles routing to sub-applications. It does simple routing
-    using regular expression matching.
+    using strict string matching.  Regular expressions are not supported.
+    Wildcard routes such as `/foo/*` are supported as a special case.
```

Comment:
As written, this implementation supports “`wildcard*`” routes even when
the wildcard is not its own path component, and so you could write a
wildcard route matching “`/foo1`”, “`/foo2`”, etc. Is this intentional,
or can we fail fast when a plugin declares a route like `/foo*`?

Separately, any thoughts as to disallowing asterisks in the middle of
routes, like `/foo/*/*` or `/bar/**/baz`, which will behave differently
than the author would expect?

These semantics should be documented on `TBPlugin.get_plugin_apps`,
whatever they end up being.

Reply:
+1 to forbidding asterisks anywhere else in the route, and to restricting them to follow a trailing `/` right now unless we have a good reason to allow generic wildcards.

Also, should we be creating a new `self.exact_routes` set for the non-wildcard routes that we can iterate over when we're doing the exact matching?  That way we won't match a request that comes in for a literal `/wildcard/*` path via the exact matching branch, which seems like a semantic error even if it's the same result in practice.

Reply:
Yep agreed throughout.  Refactored, replacing data_applications entirely with exact_routes and prefix_routes.  Added validation that there is at most one * (at the end only, and requiring a slash), and tests for that.